GH-35531: [Python] C Data Interface PyCapsule Protocol (#37797)
apache/arrow: docs/source/format/CDataInterface/PyCapsuleInterface.rst

.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

=============================
The Arrow PyCapsule Interface
=============================

Rationale
=========

The :ref:`C data interface <c-data-interface>` and
:ref:`C stream interface <c-stream-interface>` allow moving Arrow data between
different implementations of Arrow. However, these interfaces don't specify how
Python libraries should expose these structs to other libraries. Prior to this,
many libraries simply provided export to PyArrow data structures, using the
``_import_from_c`` and ``_export_from_c`` methods. However, this always required
PyArrow to be installed.

This interface allows any library to export Arrow data structures to other
libraries that understand the same protocol.

Goals
-----

* Standardize the PyCapsule objects that represent ``ArrowSchema``, ``ArrowArray``,
  and ``ArrowArrayStream``.
* Define standard methods to get these PyCapsules, so any Python objects that

>   support export via Arrow can be recognized by other libraries.

Author: Already had a link in the `PyCapsule Standard` section, but I've added another in the `Rationale` section so it's available sooner.

Non-goals
---------

* Standardizing what public APIs should be used for import. This is left up to
  individual libraries.

Comparison to DataFrame Interchange Protocol
--------------------------------------------

> :doc:`DataFrame Interchange Protocol <../dataframe.html>` is another protocol

Author: I've updated the link to that page. Used a `:doc:` directive instead though.

> :doc:`DataFrame Interchange Protocol <../dataframe.html>` is another protocol

Reviewer: (it's fine the keep the :doc: here, but in general a :ref: is more robust long term eg when restructuring documentation)

in Python that allows for the sharing of data between libraries. This protocol
is complementary to the DataFrame Interchange Protocol. Many of the objects that
implement this protocol will also implement the DataFrame Interchange Protocol.

This protocol is specific to Arrow-based data structures, while the DataFrame
Interchange Protocol allows non-Arrow data frames and arrays to be shared as well.
Because of this, these PyCapsules can support Arrow-specific features such as
nested columns.

This protocol is also much more minimal than the DataFrame Interchange Protocol.
It just handles data export, rather than defining accessors for details like
number of rows or columns.

In summary, if you are implementing this protocol, you should also consider
implementing the DataFrame Interchange Protocol.

PyCapsule Standard
==================

`PyCapsules`_ allow for a ``name`` to be associated with the capsule, allowing 
consumers to verify that the capsule contains the expected data. To make sure
Arrow structs are recognized, the following names should be used:

.. list-table::
   :widths: 25 25
   :header-rows: 1

   * - C Interface Type
     - PyCapsule Name
   * - ArrowSchema

>      - ``arrowschema``

Author: I've transformed all the names to keep with PEP 8. The only reason I did without spaces is I noticed that's how @jorisvandenbossche named `arrowarrayschema` in his ADBC PR https://github.com/apache/arrow-adbc/pull/702

>      - ``arrowschema``

Reviewer: And I don't think I had a specific reason for it, so pep8ified names is perfect

   * - ArrowArray
     - ``arrowarray``
   * - ArrowArrayStream
     - ``arrowarraystream``

Lifetime Semantics
------------------

The exported PyCapsules should have a destructor that calls the
:ref:`release callback <c-data-interface-released>`
of the Arrow struct, if it is not already null. This prevents a memory leak in
case the capsule was never passed to another consumer.

If the consumer has been passed to a consumer, the consumer should have moved
the data and marked the release callback as null, so there isn’t a risk of
releasing data the consumer is using. Read more in the C Data Interface

> specification.

Author: Done.

Because of the lifetime semantics of the underlying Arrow structs, the

> PyCapsules can only be used once.

Author: I took your suggestion.

Export Protocol
===============

The interface is three separate protocols:

* ``ArrowSchemaExportable``, which defines the ``__arrow_c_schema__`` method.
* ``ArrowArrayExportable``, which defines the ``__arrow_c_array__``. It extends
  ``ArrowSchemaExportable``, so it must also define ``__arrow_c_schema__``.
* ``ArrowStreamExportable``, which defines the ``__arrow_c_stream__``. It extends
  ``ArrowSchemaExportable``, so it must also define ``__arrow_c_schema__``.

The protocols are defined below in terms of ``typing.Protocol``. These may be
copied into a library for the purposes of static type checking, but this is not
required to implement the protocol.

.. TODO: should stream have a batch size parameter?

> .. TODO: should we have a allow_copy flag?

Author: Do folks think we need these parameters?

Reviewer: The batch size could potentially also be something that is handled in a protocol one level higher up? (like the Dataset protocol we were discussing, that sparked pushing this one forward)  
I imagine that this interface typically "just" exposes the data as they are, using the batch size as the data already exists in memory. Of course, use cases for an Arrow Stream typically involves data _not_ yet in memory, and there a batch size would be useful (but again, could also be handled by another protocol if you want to be able to control this)

            The capsule will have a name of "arrowarray".

            If requested_schema is passed, the callee should attempt to provide the
            data in the requested schema. However, this is best-effort, and the
            callee may return a PyCapsule containing an ArrowArray with a different
            schema. This parameter is useful for cases where the underlying data
            could be represented in multiple ways, and the caller has a preference
            for how it is represented. For example, some systems have a single
            integer type, but Arrow has multiple integer types with different 
            sizes and sign.

            Parameters
            ----------
            requested_schema : PyCapsule | None
                A PyCapsule containing a C ArrowSchema representation of a requested
                schema. Conversion to this schema is best-effort.

            Returns
            -------
            Tuple[PyCapsule, PyCapsule]
                A pair of PyCapsules containing a C ArrowSchema and ArrowArray,
                respectively.
            """
            ...

>     class ArrowStreamExportable(ArrowSchemaExportable, Protocol):

Author: I was thinking any object that can be exported as an array or stream also has a schema that can be exported. And I think it can be useful because just exporting the schema tells the caller what schema the data would be exported as by default if no `requested_schema` was provided.

>     class ArrowStreamExportable(ArrowSchemaExportable, Protocol):

Reviewer 2: > And for ArrowArrayExportable, we also want to expose the schema directly, I think, so this keeps it consistent

This means you could eventually pass an array or record batch to `pa.schema`, which seems weird.

>     class ArrowStreamExportable(ArrowSchemaExportable, Protocol):

Reviewer: In theory our functions like `pa.schema(..)` could reject objects that also implement `__arrow_c_array/stream__` and not only the schema one, if we want so. 

Personally I think we definitely want to give the ability to inspect the schema of an array object, before asking for the actual data (as this step could in certain cases not be zero-copy, and thus have a cost). For pyarrow objects you could check the `.type` or `.schema` attributes, and then get `__arrow_c_schema__`, but that gives again something library-specific in the middle, which we want to avoid.

>     class ArrowStreamExportable(ArrowSchemaExportable, Protocol):

Author: > This means you could eventually pass an array or record batch to pa.schema, which seems weird.

I see your point. That would be weird indeed.

I can remove this inheritance, and I'll think about whether we actually need a method to get a schema on stream / array exportables.

>     class ArrowStreamExportable(ArrowSchemaExportable, Protocol):

Reviewer 3: > Personally I think we definitely want to give the ability to inspect the schema of an array object, before asking for the actual data (as this step could in certain cases not be zero-copy, and thus have a cost).

When would this not be zero-copy? Also, an array object wouldn't have a schema ever anyways, right - it would be a separate object (or an object implementing both protocols, I suppose)

I agree we should treat them as separate objects. Note that getting the schema of a stream may involve I/O and is a falliable operation, so I think that's more reason to separate them.

>     class ArrowStreamExportable(ArrowSchemaExportable, Protocol):

Reviewer: > When would this not be zero-copy? 

For any object that does not use Arrow memory _exactly_. For example data from Velox or DuckDB right now, or pandas DataFrames, .. 
A general question might be to what extent a library should add those dunder methods if those involve a conversion step (or whether there is a general expectation that those are guaranteed to be zero-copy)

>     class ArrowStreamExportable(ArrowSchemaExportable, Protocol):

Reviewer 3: Ah, okay. Those systems could do the conversion lazily (only when you call get_next), but that's fair. (But you still don't want streams to implicitly convert to schemas, since then a system that does the conversion eagerly would do the work when you just want a schema.)

>     class ArrowStreamExportable(ArrowSchemaExportable, Protocol):

Author: FYI I've removed the inheritance here.

>     class ArrowStreamExportable(ArrowSchemaExportable, Protocol):

Reviewer: This is also something that could be handled by a protocol "higher up", like the dataset protocol we have been discussing (https://github.com/apache/arrow/issues/37504). So that way, an object could provide both a schema and a stream access, and thus a way to first inspect the schema before asking the actual stream.

.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

=============================
The Arrow PyCapsule Interface
=============================

Rationale
=========

The :ref:`C data interface <c-data-interface>` and
:ref:`C stream interface <c-stream-interface>` allow moving Arrow data between
different implementations of Arrow. However, these interfaces don't specify how
Python libraries should expose these structs to other libraries. Prior to this,
many libraries simply provided export to PyArrow data structures, using the
``_import_from_c`` and ``_export_from_c`` methods. However, this always required

> PyArrow to be installed.

Reviewer 2: You could also mention that any misuse of those APIs, or any exception happening between exporting and importing, could result in a memory leak.

> PyArrow to be installed.

Author: ```suggestion
PyArrow to be installed. In addition, those APIs could cause memory leaks if
handled improperly.
```

This interface allows any library to export Arrow data structures to other
libraries that understand the same protocol.

Goals
-----

* Standardize the `PyCapsule`_ objects that represent ``ArrowSchema``, ``ArrowArray``,
  and ``ArrowArrayStream``.

> * Define standard methods to get these capsule objects, so any Python objects that
>   support export via Arrow can be recognized by other libraries.

Reviewer 2: Perhaps we can a bit more precise here.
```suggestion
* Define standard methods that export Arrow data into such capsule objects,
  so that any Python library wanting to accept Arrow data as input can call the
  corresponding method instead of hardcoding support for specific Arrow
  producers.
```

Non-goals
---------

* Standardizing what public APIs should be used for import. This is left up to
  individual libraries.

> Comparison to DataFrame Interchange Protocol

Reviewer 2: I would move the "comparison" sections towards the end, since they are not required for understanding and implementing the spec.

--------------------------------------------

`The DataFrame Interchange Protocol <https://data-apis.org/dataframe-protocol/latest/>`_
is another protocol in Python that allows for the sharing of data between libraries.
This protocol is complementary to the DataFrame Interchange Protocol. Many of
the objects that implement this protocol will also implement the DataFrame
Interchange Protocol.

This protocol is specific to Arrow-based data structures, while the DataFrame
Interchange Protocol allows non-Arrow data frames and arrays to be shared as well.
Because of this, these PyCapsules can support Arrow-specific features such as
nested columns.

This protocol is also much more minimal than the DataFrame Interchange Protocol.
It just handles data export, rather than defining accessors for details like
number of rows or columns.

In summary, if you are implementing this protocol, you should also consider
implementing the DataFrame Interchange Protocol.

> Comparison to __arrow_array__ protocol
> --------------------------------------

Reviewer 2: ```suggestion
Comparison to ``__arrow_array__`` protocol
------------------------------------------
```

> The ``__arrow_array__`` protocol is a dunder method that defines how PyArrow

Reviewer 2: Link to the relevant section in the Python docs?

> should import an object as an Arrow array. Unlike this protocol, it does
> specific to PyArrow and isn't used by other libraries. It is also limited to

Reviewer: ```suggestion
should import an object as an Arrow array. Unlike this protocol, it is
specific to PyArrow and isn't used by other libraries. It is also limited to
```

arrays and does not support schemas, tabular structures, or streams.

PyCapsule Standard
==================

`PyCapsule`_ allows for a ``name`` to be associated with the capsule, allowing 

> consumers to verify that the capsule contains the expected data. To make sure

Reviewer 2: ```suggestion
consumers to verify that the capsule contains the expected kind of data. To make sure
```

> Arrow structs are recognized, the following names should be used:

Reviewer 2: ```suggestion
Arrow structs are recognized, the following names must be used:
```

.. list-table::
   :widths: 25 25
   :header-rows: 1

   * - C Interface Type
     - PyCapsule Name
   * - ArrowSchema
     - ``arrow_schema``
   * - ArrowArray
     - ``arrow_array``
   * - ArrowArrayStream
     - ``arrow_array_stream``

Lifetime Semantics
------------------

The exported PyCapsules should have a destructor that calls the
:ref:`release callback <c-data-interface-released>`
of the Arrow struct, if it is not already null. This prevents a memory leak in
case the capsule was never passed to another consumer.

If the capsule has been passed to a consumer, the consumer should have moved
the data and marked the release callback as null, so there isn’t a risk of
releasing data the consumer is using.
:ref:`Read more in the C Data Interface specification <c-data-interface-released>`.

Just like in the C Data Interface, the PyCapsule objects defined here can only
be consumed once.

Export Protocol
===============

> The interface is three separate protocols:

Reviewer 2: "consists in" or "consists of" perhaps?

> * ``ArrowSchemaExportable``, which defines the ``__arrow_c_schema__`` method.

Reviewer 2: Since the protocol classes themselves are optional and do not have any runtime effect, perhaps this should be restructured to first describe the spec itself (the methods and what they do), and then another subsection for optional typing support?

Reviewer 2: For example this is how Python describes the builtin special methods such as `__len__`:
https://docs.python.org/3/reference/datamodel.html#emulating-container-types

Author: Thanks. I've refactored to do that. It's quite nice in that I can now reference the methods elsewhere.

> * ``ArrowArrayExportable``, which defines the ``__arrow_c_array__``.
> * ``ArrowStreamExportable``, which defines the ``__arrow_c_stream__``.

Reviewer: ```suggestion
* ``ArrowArrayExportable``, which defines the ``__arrow_c_array__`` method.
* ``ArrowStreamExportable``, which defines the ``__arrow_c_stream__`` method.
```

(or leave the "the")

            for how it is represented. For example, some systems have a single
            integer type, but Arrow has multiple integer types with different 
            sizes and sign.

            Parameters
            ----------
            requested_schema : PyCapsule | None
                A PyCapsule containing a C ArrowSchema representation of a requested
                schema. Conversion to this schema is best-effort.

            Returns
            -------
            PyCapsule
                A PyCapsule containing a C ArrowArrayStream representation of the
                object.
            """
            ...

Schema Requests
---------------

In some cases, there might be multiple possible Arrow representations of the
same data. For example, a library might have a single integer type, but Arrow
has multiple integer types with different sizes and sign. As another example,
Arrow has several possible encodings for an array of strings: 32-bit offsets,
64-bit offsets, string view, and dictionary-encoded. A sequence of strings could
export to any one of these Arrow representations.

In order to allow the caller to request a specific representation, the

> ``__arrow_c_array__`` and ``__arrow_c_stream__`` methods may take an optional

Reviewer 2: ```suggestion
``__arrow_c_array__`` and ``__arrow_c_stream__`` methods take an optional
```

``requested_schema`` parameter. This parameter is a PyCapsule containing an
``ArrowSchema``.

The callee should attempt to provide the data in the requested schema. However,
if the callee cannot provide the data in the requested schema, they may return
with the same schema as if ``None`` were passed to ``requested_schema``.

If the caller requests a schema that is not compatible with the data,
say requesting a schema with a different number of fields, the callee should

> raise an exception. The requested schema mechanism is only meant to negotiate

Reviewer 2: ```suggestion
raise an exception, preferably ``TypeError`` or a subclass thereof.
The requested schema mechanism is only meant to negotiate
```

Reviewer: "preferably" is of course not strong, so maybe fine to keep, but I am not sure we generally should expect a single class of error. There are multiple reasons why a certain cast is not possible, for example because the types are completely incompatible (TypeError), or because a specific value in the array cannot be cast (ValueError).

Reviewer 2: Hmm, that's true. I withdraw my suggestion then.

between different representations of the same data and not to allow arbitrary
schema transformations.

.. _PyCapsule: https://docs.python.org/3/c-api/capsule.html

Examples
========

Create a PyCapsule
------------------

To create a PyCapsule, use the `PyCapsule_New <https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_New>`_
function. The function must be passed a destructor function that will be called
to release the data the capsule points to. It must first call the release
callback if it is not null, then free the struct.

Below is the code to create a PyCapsule for an ``ArrowSchema``. The code for
``ArrowArray`` and ``ArrowArrayStream`` is similar.

> .. tab-set::

Reviewer 2: The examples show a `free` without a corresponding `malloc`, which may be confusing and error-prone. I think you should show both.

(also discuss that the choice of malloc/free is producer-specific and other allocation means can be used if desired, as long as the capsule's destructor calls the appropriate deallocation function)

Author: Changed to show the explicit malloc.

   .. tab-item:: C

      .. code-block:: c

         #include <Python.h>

         void ReleaseArrowSchemaPyCapsule(PyObject* capsule) {
             ArrowSchema* schema =
                 (ArrowSchema*)PyCapsule_GetPointer(capsule, "arrow_schema");
             if (schema->release != NULL) {
                 schema->release(schema);
             }
             free(schema);
         }
         
         PyObject* MakeArrowSchemaPyCapsule(ArrowSchema* schema) {
           return PyCapsule_New(schema, "arrow_schema", ReleaseArrowSchemaPyCapsule);
         }

   .. tab-item:: Cython

      .. code-block:: cython

>          import cpython

Reviewer 2: Does this work or does this need to be `cimport cpython`?

Reviewer: Yes, that needs to be `cimport`

Author: Thanks. I've now fixed the examples and tested they can compile successfully.

         cdef void release_arrow_schema_py_capsule(object schema_capsule):
             cdef ArrowSchema* schema = <ArrowSchema*>cpython.PyCapsule_GetPointer(
                 schema_capsule, 'arrow_schema'
             )
             if schema.release != NULL:
                 schema.release(schema)
         
             free(schema)
         
         cdef object make_arrow_schema_py_capsule(ArrowSchema* schema):
             return cpython.PyCapsule_New(
                 <void*>schema, 'arrow_schema', release_arrow_schema_py_capsule
             )

Consume a PyCapsule
-------------------

To consume a PyCapsule, use the `PyCapsule_GetPointer <https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_GetPointer>`_ function

> to get the pointer to the underlying struct. Import the struct using the your
> systems Arrow C Data Interface import function. Only after that should the

Reviewer: ```suggestion
to get the pointer to the underlying struct. Import the struct using your
system's Arrow C Data Interface import function. Only after that should the
```

capsule be freed.

The below example shows how to consume a PyCapsule for an ``ArrowSchema``. The
code for ``ArrowArray`` and ``ArrowArrayStream`` is similar.

.. tab-set::

   .. tab-item:: C

      .. code-block:: c

         #include <Python.h>
         
         // If the capsule is not an ArrowSchema, will return NULL.
         ArrowSchema* GetArrowSchemaPyCapsule(PyObject* capsule) {
           return PyCapsule_GetPointer(schema, "arrow_schema");
         }

   .. tab-item:: Cython

      .. code-block:: cython

         import cpython
        
         cdef object get_arrow_schema_py_capsule(PyObject* capsule):
             return cpython.PyCapsule_GetPointer(capsule, 'arrow_schema')

> Backwards Compatibility with PyArrow

Reviewer 2: I always misremember which is "forwards" and which is "backwards" compatibility...

------------------------------------

> When interacting with PyArrow, the PyCapsule interface will be preferred over

Reviewer 2: ```suggestion
When interacting with PyArrow, the PyCapsule interface should be preferred over
```

       array_import_ptr = make_array_import_ptr()
       schema_import_ptr = make_schema_import_ptr()
       arr._export_to_c(array_import_ptr, schema_import_ptr)
       return import_c_data(array_import_ptr, schema_import_ptr)

You can rewrite this method to support both PyArrow and other libraries that
implement the PyCapsule interface:

.. code-block:: python

   # NEW METHOD
   def from_arrow(arr)
       if hasattr(arr, "__arrow_c_array__"):
            schema_ptr, array_ptr = arr.__arrow_c_array__()
            return import_c_capsule_data(schema_ptr, array_ptr)
       elif isinstance(arr, pa.Array):
            # Deprecated method, used for older versions of PyArrow
            array_import_ptr = make_array_import_ptr()
            schema_import_ptr = make_schema_import_ptr()
            arr._export_to_c(array_import_ptr, schema_import_ptr)
            return import_c_data(array_import_ptr, schema_import_ptr)
       else:
           raise TypeError(f"Cannot import {type(arr)} as Arrow array data.")

You may also wish to accept objects implementing the protocol in your
constructors. For example, in PyArrow, the :py:func:`array` and :py:func:`record_batch`

> constructor accepts any object that implements the ``ArrowArrayExportable``

Reviewer 2: ```suggestion
constructors accept any object that implements the ``ArrowArrayExportable``
```

protocol. Similarly, the PyArrow's :py:func:`schema` constructor accepts any object

> that implements the ``ArrowSchemaExportable`` protocol.

Reviewer 2: ```suggestion
that implements the ``__arrow_c_schema__`` method.
```

Now if your library has an export to PyArrow function, such as:

.. code-block:: python

   # OLD METHOD
   def to_arrow(self) -> pa.Array:
       array_export_ptr = make_array_export_ptr()
       schema_export_ptr = make_schema_export_ptr()
       self.export_c_data(array_export_ptr, schema_export_ptr)
       return pa.Array._import_from_c(array_export_ptr, schema_export_ptr)

You can implement the protocol instead and encourage users to prefer directly
passing your object to the :py:func:`array` constructor, which accepts any object

> that implements the protocol:

Reviewer: I am not sure we should recommend deprecating such explicit methods? (personally I am not sure I would do that if I were maintaining such a library, it's also nice (and more discoverable) to have an explicit method for it)

Reviewer 2: Indeed, I would leave it to third-party libraries to make such an API decision.

Author: I've removed this.

.. code-block:: python

  import warnings

  # NEW METHOD
  def to_arrow(self) -> pa.Array:

>       if hasattr(pa.Array, "__arrow_c_array__"):

Reviewer 2: You might add a comment explaining why the `hasattr` call, because it was not immediately obvious for me.

         
         cdef object export_arrow_schema_py_capsule():
             cdef ArrowSchema* schema = <ArrowSchema*>malloc(sizeof(ArrowSchema))
             # It's recommended to immediately wrap the struct in a capsule, so
             # if subsequent lines raise an exception memory will not be leaked.
             schema.release = NULL
             capsule = cpython.PyCapsule_New(
                 <void*>schema, 'arrow_schema', release_arrow_schema_py_capsule
             )
             # Fill in ArrowSchema fields:
             # schema.format = ...
             # ...
             return capsule

Consume a PyCapsule
-------------------

To consume a PyCapsule, use the `PyCapsule_GetPointer <https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_GetPointer>`_ function
to get the pointer to the underlying struct. Import the struct using your
system's Arrow C Data Interface import function. Only after that should the
capsule be freed.

The below example shows how to consume a PyCapsule for an ``ArrowSchema``. The
code for ``ArrowArray`` and ``ArrowArrayStream`` is similar.

.. tab-set::

   .. tab-item:: C

      .. code-block:: c

         #include <Python.h>
         

>          // If the capsule is not an ArrowSchema, will return NULL.

Reviewer 2: ```suggestion
         // If the capsule is not an ArrowSchema, will return NULL and set an exception.
```

         struct ArrowSchema* GetArrowSchemaPyCapsule(PyObject* capsule) {
           return PyCapsule_GetPointer(capsule, "arrow_schema");
         }

   .. tab-item:: Cython

      .. code-block:: cython

         cimport cpython
        

>          cdef ArrowSchema* get_arrow_schema_py_capsule(object capsule):

Reviewer 2: ```suggestion
         cdef ArrowSchema* get_arrow_schema_py_capsule(object capsule) except NULL:
```
